try is a keyword in rust 2018 edition

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried running gccrs against libcore 1.19 and got:

```
$ /opt/gccrs/bin/gccrs -g -O0 -frust-crate=core -c libcore/lib.rs -o libcore.o
libcore/macros.rs:329:14: error: expecting ‘identifier’ but ‘try’ found
  329 | macro_rules! try {
      |              ^
libcore/macros.rs:329:14: error: failed to parse item in crate
rust1: internal compiler error: Segmentation fault
0x1b7e685 crash_signal
        ../../gccrs/gcc/toplev.cc:322
0x7f40a6abf08f ???
        /build/glibc-SzIz7B/glibc-2.31/signal/../sysdeps/unix/sysv/linux/x86_64/sigaction.c:0
0x1032fa8 Rust::AST::Type::clone_type() const
        ../../gccrs/gcc/rust/ast/rust-ast.h:1108
0x11602c7 Rust::AST::SingleASTNode::SingleASTNode(Rust::AST::SingleASTNode const&)
        ../../gccrs/gcc/rust/ast/rust-ast.h:1636
0x11ae7f4 void std::_Construct<Rust::AST::SingleASTNode, Rust::AST::SingleASTNode const&>(Rust::AST::SingleASTNode*, Rust::AST::SingleASTNode const&)
        /usr/include/c++/9/bits/stl_construct.h:75
0x11ad393 Rust::AST::SingleASTNode* std::__uninitialized_copy<false>::__uninit_copy<Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode*>(Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode*)
        /usr/include/c++/9/bits/stl_uninitialized.h:83
0x11a7ca9 Rust::AST::SingleASTNode* std::uninitialized_copy<Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode*>(Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode*)
        /usr/include/c++/9/bits/stl_uninitialized.h:140
0x1195809 Rust::AST::SingleASTNode* std::__uninitialized_copy_a<Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode*, Rust::AST::SingleASTNode>(Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode*, std::allocator<Rust::AST::SingleASTNode>&)
        /usr/include/c++/9/bits/stl_uninitialized.h:307
0x1181196 void std::vector<Rust::AST::SingleASTNode, std::allocator<Rust::AST::SingleASTNode> >::_M_range_initialize<Rust::AST::SingleASTNode const*>(Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode const*, std::forward_iterator_tag)
        /usr/include/c++/9/bits/stl_vector.h:1582
0x116ba5d std::vector<Rust::AST::SingleASTNode, std::allocator<Rust::AST::SingleASTNode> >::vector(std::initializer_list<Rust::AST::SingleASTNode>, std::allocator<Rust::AST::SingleASTNode> const&)
        /usr/include/c++/9/bits/stl_vector.h:626
0x115dbee transcribe_type
        ../../gccrs/gcc/rust/expand/rust-macro-expand.cc:881
0x115ddf1 transcribe_context
        ../../gccrs/gcc/rust/expand/rust-macro-expand.cc:933
0x115e230 Rust::MacroExpander::transcribe_rule(Rust::AST::MacroRule&, Rust::AST::DelimTokenTree&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, Rust::MatchedFragmentContainer, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, Rust::MatchedFragmentContainer> > >&, bool, Rust::MacroExpander::ContextType)
        ../../gccrs/gcc/rust/expand/rust-macro-expand.cc:1013
0x115ae15 Rust::MacroExpander::expand_decl_macro(Location, Rust::AST::MacroInvocData&, Rust::AST::MacroRulesDefinition&, bool)
        ../../gccrs/gcc/rust/expand/rust-macro-expand.cc:105
0x115b3a1 Rust::MacroExpander::expand_invoc_semi(Rust::AST::MacroInvocation&)
        ../../gccrs/gcc/rust/expand/rust-macro-expand.cc:203
0x11b181c Rust::AttrVisitor::visit(Rust::AST::MacroInvocation&)
        ../../gccrs/gcc/rust/expand/rust-attribute-visitor.cc:406
0x1030611 Rust::AST::MacroInvocation::accept_vis(Rust::AST::ASTVisitor&)
        ../../gccrs/gcc/rust/ast/rust-ast-full-test.cc:5516
0x11603e4 Rust::AST::SingleASTNode::accept_vis(Rust::AST::ASTVisitor&)
        ../../gccrs/gcc/rust/ast/rust-ast.h:1766
0x1161474 Rust::MacroExpander::take_expanded_fragment(Rust::AST::ASTVisitor&)
        ../../gccrs/gcc/rust/expand/rust-macro-expand.h:301
0x11bba1b Rust::AttrVisitor::maybe_expand_type(std::unique_ptr<Rust::AST::Type, std::default_delete<Rust::AST::Type> >&)
        ../../gccrs/gcc/rust/expand/rust-attribute-visitor.cc:3444
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.

```

### Meta

- What version of Rust GCC were you using, git sha if possible. fb1b2a6bf5eed9dd9405dca9b9c895f48509875e

--- a/gcc/rust/lex/rust-lex.cc
+++ b/gcc/rust/lex/rust-lex.cc
@@ -21,6 +21,7 @@
 #include "rust-system.h"      // for rust_assert and rust_unreachable
 #include "rust-diagnostics.h" // for rust_error_at
 #include "rust-linemap.h"
+#include "rust-session-manager.h"
 #include "safe-ctype.h"
 
 namespace Rust {
@@ -228,10 +229,24 @@ Lexer::classify_keyword (const std::string &str)
 
   if (idx == last || str != *idx)
     return IDENTIFIER;
-  else
-    return keyword_keys[idx - keyword_index];
 
   // TODO: possibly replace this x-macro system with something like hash map?
+
+  // We now have the expected token ID of the reserved keyword. However, some
+  // keywords are reserved starting in certain editions. For example, `try` is
+  // only a reserved keyword in editions >=2018. The language might gain new
+  // reserved keywords in the future.
+  //
+  // https://doc.rust-lang.org/reference/keywords.html#reserved-keywords
+  auto id = keyword_keys[idx - keyword_index];
+
+  // `try` is not a reserved keyword before 2018
+  if (Session::get_instance ().options.get_edition ()
+	== CompileOptions::Edition::E2015
+      && id == TRY)
+    return IDENTIFIER;
+
+  return id;
 }
 
 TokenPtr

--- a/gcc/rust/rust-session-manager.h
+++ b/gcc/rust/rust-session-manager.h
@@ -191,13 +191,13 @@ struct CompileOptions
   bool proc_macro = false;
   std::string metadata_output_path;
 
-  enum Edition
+  enum class Edition
   {
     E2015 = 0,
     E2018,
     E2021,
   } edition
-    = E2015;
+    = Edition::E2015;
 
   bool dump_option_enabled (DumpOption option) const
   {
@@ -238,6 +238,8 @@ struct CompileOptions
   {
     edition = static_cast<Edition> (raw_edition);
   }
+
+  const Edition &get_edition () { return edition; }
 
   void set_metadata_output (const std::string &path)
   {

--- a/gcc/testsuite/rust/compile/macro-issue1395-2.rs
+++ b/gcc/testsuite/rust/compile/macro-issue1395-2.rs
@@ -1,0 +1,7 @@
+// { dg-additional-options "-frust-edition=2018" }
+
+macro_rules! try {
+    // { dg-error "expecting .identifier. but .try. found" "" { target *-*-* } .-1 }
+    // { dg-error "failed to parse item in crate" "" { target *-*-* } .-2 }
+    () => {};
+}

--- a/gcc/testsuite/rust/compile/macro-issue1395.rs
+++ b/gcc/testsuite/rust/compile/macro-issue1395.rs
@@ -1,0 +1,5 @@
+// Default edition is 2015 - this is valid
+
+macro_rules! try {
+    () => {};
+}

